### ISSUE TITLE
vRO plugin - vRA 7.4 - properties issues

### ISSUE THREAD

**Issue by 5280tunage** [puppetlabs/puppet-vro-starter_content#36]
Hi there, this may or may not be the right place for this, so i apologize if it's not.  Have deployed the Puppet plugin for vRO/vRA, in a new environment, and for the most part it's working.  Quick note, I'm using vRA 7.4 and the latest vRO version as well.  I'm also using PE version 2017.3.5.

The plug-in itself appears to be working fine, however the data look ups don't seem to be working properly.  Specifically, when creating a new blueprint for a traditional linux VM, the Puppet module works fine until i'm trying to select the Puppet role.  Specifically, selecting the puppet master and environment appears to work just fine, it dynamically populates the environments, but when i go to select the role, the roles do not appear.

I've ensured that the vro plugin user should have appropriate permissions, even assigning that user to the administrator role to test.  Any thoughts?  I'm finding absolutely no errors or events in vRO that help me troubleshoot the puppet code in vRO.  Also don't see any errors on the Puppet side either.

In this case I am trying to use the starter role that was created, role::linux_webserver

![image](https://user-images.githubusercontent.com/22239585/39480087-64647c64-4d24-11e8-9e45-2ae500106a39.png)

**Comment by jpadams** [comment]
@jhedman2  Do you use the standard control-repo format and locations of /etc/puppetlabs/code/environments/dev/site/role/manifests/linux_webserver.pp?

It sounds like you're using the https://github.com/puppetlabs/puppet-vro-starter_content repo in its standard form, set up by the vra_nc_setup.sh script, so the answer is likely "yes" to the above, unless you're using the content, but putting into a different file system structure.

**Comment by 5280tunage** [comment]
@jpadams Correct, at this time I am using the standards, and yes, I'm starting from scratch using the starter content referenced previously.  Outside of that, I haven't modified any Puppet configurations beyond default yet, I'm somewhat new to it, just trying to get it to work before i mess things up :)

**Comment by jpadams** [comment]
@jhedman2 Assuming you're using a very recent pull of the starter content, but wanted to confirm.

One thing you could try would be to set up a new Puppet plugin endpoint or edit your existing `homepuppet01` endpoint to use the Puppet master's `root` user instead. If things work like that, I'd suspect something around the config of the `vro-plugin-user` and/or the perms/ownership of the /etc/puppetlabs/code/*directories (codedir)

The starter content employs the `vra_puppet_plugin_prep` module to set things up on the master and we use this code to set up the `vro-plugin-user`: https://github.com/puppetlabs/puppet-vra_puppet_plugin_prep/blob/master/manifests/init.pp#L64-L70
In this example we put that user in the `pe-puppet` group to ensure that it can read code in the codedir. Good to confirm that your user is set up like that for now.

We have some updates coming to make `sudo` work more reliably out of the box, but hoping you can test these couple things to help isolate the issue.

**Comment by 5280tunage** [comment]
Thank you again for the fast response.  I have verified that the vro-plugin-user account is a member of the pe-puppet group.  I did notice that in the directory structure, the pe-puppet group only appears to have access permissions up to the /etc/puppetlabs/code level.  beyond that, only root has permissions.

With that said, I did enable ssh login for the root user, and tested that in the plugin for the endpoint, and it still isn't showing the roles.

![image](https://user-images.githubusercontent.com/22239585/39484484-2e730b52-4d33-11e8-8333-19cc194fc681.png)

Thanks again.  happy to try other items, enable any sort of verbose logging, or add more data if it helps.  I am on Ubuntu 16.04 btw.

**Comment by jpadams** [comment]
@jhedman2 Thanks for testing that. Ideally, everything under `/etc/puppetlabs/code/*` should be owned by `pe-puppet`. I may have to adjust the setup script to ensure that. For now, could you run something like:

```
chown -R pe-puppet:pe-puppet /etc/puppetlabs/code/*
```

If you're handy with vRO, you could run the action that vRA uses within vRO directly by creating a test workflow. Ping me at jeremy@puppet.com if you have time to do a screen share.

**Comment by 5280tunage** [comment]
sent you a quick note, please let me know if you didn't receive it by chance.

**Comment by jpadams** [comment]
@jhedman2 this might be due to a change in the puppet-strings gem. On your master, could you remove the gem and see if things resolve. You might get role classes without descriptions.

```
puppet resource package puppet-strings ensure=absent
```

If so, then there's a tweak needed to how puppet-strings is called. That will be out in next release of plugin.

**Comment by 5280tunage** [comment]
Man, this is a head scratch-er, the roles still won't populate in the vRO design UI.  Specifically, when selecting the builtin Dev or Production environments in the blueprint drop down but the roles won't populate.  So for example, here's the list of roles - 
![image](https://user-images.githubusercontent.com/22239585/40379466-6581286e-5db3-11e8-9bdb-ad50f8d910e9.png)

and here's what the blueprint designer shows.
![image](https://user-images.githubusercontent.com/22239585/40379498-7b645516-5db3-11e8-9ebe-a4ffbceed230.png)

**Comment by 5280tunage** [comment]
Oh, and i did update the master config as you showed above.  Does that require a service/master restart to take effect?

**Comment by EclecticCurmudgeon** [comment]
Was there any resolution to this?
I'm running into the same issue.

**Comment by 5280tunage** [comment]
Actually yes, my issue turned out to be that I was using the wrong version
of the plug-in.  Basically we ended up having to go remove the vro
endpoint, then we remove the plug in via the vro console, then we were able
to install the new version of the plugin, set up the endpoint and
everything started to work properly. If you already have vro workflows that
utilize it you don't have to change anything they should begin working.  In
my case I simply installed the new plug-in and endpoint using the exact
same name as the previous, and this made sure the workflows continue to
work.

**Comment by EclecticCurmudgeon** [comment]
Good the hear, what version of the plugin are you running, I'm running 3.0.0

**Comment by 5280tunage** [comment]
I believe 3.1 was just released, that's what I'm running.

**Comment by 5280tunage** [comment]
So the VMware solutions exchange website is kind of confusing, if you got the resources page at https://marketplace.vmware.com/vsx/solutions/puppet-plugin-for-vrealize-automation#resources, instead of looking on the left for the latest download, you have to click the "Try" button on the right, that will download the latest version which is 3.1 right now.

Hope that helps.

**Comment by EclecticCurmudgeon** [comment]
Thanks for the help, I uninstalled the old plugin installed 3.1 and still not getting anything, I'm going to try a few things in the rest of this thread again see if I can find what works.

**Comment by jpadams** [comment]
@Mossmanxx Now that you have the 3.1 plugin, have you re-registered your Puppet endpoint in vRA? I'd delete the old Puppet endpoint and add a new one. And to be clear, you've applied this starter content to your Puppet Enterprise (not open source) master, right? Version of PE? 2017.3.x or 2018.x should be good.

**Comment by EclecticCurmudgeon** [comment]
Okay I did some testing, we aren't using the starter content outright because we already have a production PE master running. When I created the path above.
/etc/puppetlabs/code/environments/dev/site/role/manifests/linux_webserver.pp

It does pull in the Role, so my next challenge it finding a way to have it so the path it's looking for is /etc/puppetslabs/code/environments/dev/modules/role/manifests

**Comment by mjbain151** [comment]
Was a solution found for configuring the path the plugin searches for roles?  I am in exactly the same position as Mossmanxx, I already have a working puppet environment with code manager running ,  so I couldn't use the vro-stater-content.  I got all to work with the exception of roles which I have in /etc/puppetslabs/code/environments/dev/modules/roles/manifests.  I can see that the plugin is searching /etc/puppetlabs/code/environments/dev/site/role/manifests.  How can that be changed?
